### PR TITLE
Union Pooler Experiment 1

### PR DESCRIPTION
**Pull request by rmccall84**
Fixes https://github.com/numenta/nupic.research/issues/157

### CODE REVIEW

**Comment by scottpurdy** [comment]
Code looks pretty clean. There is a lot here so I would recommend breaking it up and getting code review asap. For instance, the union pooler class could be its own PR.

And you should include unit tests for the different pieces.

**Comment by rmccall84** [comment]
@scottpurdy @chetan51 I'm going to take pieces of this PR and merge them in separate PRs:

https://github.com/numenta/nupic.research/pull/161
https://github.com/numenta/nupic.research/pull/162

Once these two are merged, I merge master into this branch. This PR will basically add the first Union Pooler experiment.

**Comment by rmccall84** [comment]
@chetan51 could you please review this. if there's major issues I'll spin them off into new PRs. right now it'll help a lot to get this merged. thx

File: union_pooling/tests/unit/union_pooler_test.py [hunk 1]
```diff
@@ -35,29 +41,74 @@ def setUp(self):
...
+    expected = [0.0, 0.0, 0.0, 0.456, 0.789]
+
+    result = self.unionPooler._addToPoolingActivation(activeCells, overlaps)
+
+    self.assertEquals(len(expected), len(result))
+    for i in xrange(len(expected)):
+      self.assertAlmostEquals(expected[i], result[i], places=6)
+
+
+  def testAddToPoolingActivation2(self):
```
Comment (chetan51) [inline union_pooling/tests/unit/union_pooler_test.py]: Please make this test name clearer (not sure what `2` means)

File: union_pooling/tests/unit/union_pooler_test.py [hunk 2]
```diff
@@ -35,29 +41,74 @@ def setUp(self):
...
+    expected = [0.0, 1.0, 2.0, 3.456, 4.789]
+
+    result = self.unionPooler._addToPoolingActivation(activeCells, overlaps)
+
+    self.assertEquals(len(expected), len(result))
+    for i in xrange(len(expected)):
+      self.assertAlmostEquals(expected[i], result[i], places=6)
+
+
+  def testGetMostActiveCellsCornerCase(self):
```
Comment (chetan51) [inline union_pooling/tests/unit/union_pooler_test.py]: Please make this test name clearer (what is the corner case?)

File: union_pooling/union_pooling/experiments/noisy_sequences/experiment.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+Generate sequences from an alphabet.
+Train on sequences.
+Test phase: Input sequence pattern by pattern. Sequence-to-sequence
+progression is randomly selected. At each step there is a chance that the
+next pattern in the sequence is not shown. Specifically the following
+perturbations may occur:
+  1) random Jump to another sequence
+  2) substitution of some other pattern for the normal expected pattern
+  3) skipping expected pattern and presenting next pattern in sequence
+  4) addition of some other pattern putting off expected pattern one time step
```
Comment (chetan51) [inline union_pooling/union_pooling/experiments/noisy_sequences/experiment.py]: What do you hope to see as a result?

File: union_pooling/union_pooling/experiments/union_sdr_overlap/experiment.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+  fileName = outputPrefix + "-union_sdr_trace.csv"
+  filePath = os.path.join(outputDir, fileName)
+  with open(filePath, "wb") as outputFile:
+    csvWriter = csv.writer(outputFile)
+    unionSdrTrace = experiment.up._mmTraces["activeCells"].data
+    rows = [list(unionSdr) for unionSdr in unionSdrTrace]
+    csvWriter.writerows(rows)
+    outputFile.flush()
+
+  # Output metrics
```
Comment (chetan51) [inline union_pooling/union_pooling/experiments/union_sdr_overlap/experiment.py]: Enable or remove commented-out code

**Comment by chetan51** [comment]
Finished review.

**Comment by rmccall84** [comment]
@chetan51 many thanks for the useful feedback
